README.md
# ORCID Utils


![Node.js CI](https://github.com/alexswilliams/orcid-utils/workflows/Node.js%20CI/badge.svg)

Library to manipulate [ORCID](https://orcid.org/) identifiers.
Allows validation of identifiers based on their check-digit,
and to format a provided identifier into a number of potentially
useful formats. 

Designed to be used as either a `require`-able module, or as 
a standalone object that can be imported into a normal HTML page.

## Node Installation

`npm install orcid-utils --save`

## Example Node Usage

```javascript
const ORCID = require('orcid-utils');
const orcidNumber = '0000-0000-0000-0001';

if (ORCID.isValid(orcidNumber)) {

    console.log(ORCID.toDashFormat(orcidNumber));
    // 0000-0000-0000-0001
    
    console.log(ORCID.toNoDashFormat(orcidNumber));
    // 0000000000000001
    
    
    console.log(ORCID.toUriWithoutProtocol(orcidNumber));
    // orcid.org/0000-0000-0000-0001
    
    console.log(ORCID.toUriWithProtocol(orcidNumber));
    // https://orcid.org/0000-0000-0000-0001
    

    // toUriWithProtocol also takes a boolean parameter that
    // specifies whether the URI should use https or not.
    
    console.log(ORCID.toUriWithProtocol(orcidNumber, true));
    // https://orcid.org/0000-0000-0000-0001    
    console.log(ORCID.toUriWithProtocol(orcidNumber, false));
    // http://orcid.org/0000-0000-0000-0001


    // Validation can also be done as a side effect:

    ORCID.validate(orcidNumber)
    // passes silently
    ORCID.validate('xxxxx')
    // throws

}
```

## Node Testing

Testing is performed using the `jest` package, and is aliased
to the `test` script.

`npm test`

## Example Browser Usage

```html
<html>
<head>
<script src="lib/orcid.min.js"></script>
</head>

<body>
<input type="text" id="orcidTextInput">
<input type="button" onClick="validateAndLog();">

    <script>
      var validateAndLog = function validateAndLog() {
        var orcidNumber = document.getElementById('orcidTextInput').value;

        if (ORCID.isValid(orcidNumber)) {
          console.log(ORCID.toDashFormat(orcidNumber));
          // 0000-0000-0000-0001
          
          // All the other examples from above work here too!
        } else {
          console.log('Not Valid');
        }

      };
    </script>

</body>
</html>
```

## How to Contribution

Please do report any issues you have with this to the project's
[GitHub Issue Tracker](https://github.com/alexswilliams/orcid-utils/issues).

If you wish to make any modifications or amendments, I'm happy to
review pull requests :).



CHANGELOG.md
## 1.0.6

- Possible breaking change: non-boolean flags will now be coerced rathern than rejected outright
- Switched to minifying with esbuild
- Now targeting, as minimum supported versions:

  - Chrome 90
  - Edge 91
  - Firefox 89
  - iOS 14.0
  - Safari 14
  - Node 12

- Rewrote code for size optimisation
- Smaller bundle size (1.81 KB -> 914 B)
- Security updates to dev dependencies

## 1.0.5 - 2020-12-10

- Bundle minification changed (1.82 KB -> 1.81 KB)
- Add support for sourcemaps
- Security updates to dev dependencies
- Remove babel; swap to google closure compiler
- Drop support for IE10

## 1.0.4 - 2019-09-25

- New function: `validate('...')` - throws whenever `isValid('...')` would have returned `false`
- Security updates to dev dependencies
- Switch form nodeunit to jest

## 1.0.3 - 2019-05-01

- Switch to using a Makefile for bulding
- Target >2% in browserslist
- Security updates to dev dependencies

## 1.0.2 - 2018-09-17

- Security updates to dev dependencies

## 1.0.1 - 2017-10-28

- Add explicit IE 10 support
- Update source to ES6 and add transpile step for compatibility

## 1.0.0 - 2017-10-28

Initial Release


test/orcid.js
"use strict";

const ORCID = require('../lib/orcid');

exports.toDashFormat = {
  non_string_inputs_throw_TypeError: function(test) {
    test.throws(function() { ORCID.toDashFormat(10); });
    test.throws(function() { ORCID.toDashFormat({}); });
    test.throws(function() { ORCID.toDashFormat([]); });
    test.throws(function() { ORCID.toDashFormat(NaN); });
    test.throws(function() { ORCID.toDashFormat(null); });
    test.throws(function() { ORCID.toDashFormat(undefined); });
    test.throws(function() { ORCID.toDashFormat(new Error()); });
    test.throws(function() { ORCID.toDashFormat(function() {}); });
    test.throws(function() { ORCID.toDashFormat(Symbol()); });
    test.throws(function() { ORCID.toDashFormat(/./); });
    test.done();
  },

  valid_ORCIDs_are_converted_to_standalone_dashed_format: function(test) {
    test.strictEqual(ORCID.toDashFormat('0000-0000-0000-0001'),
                     '0000-0000-0000-0001');
    test.strictEqual(ORCID.toDashFormat('0001-2345-6789-012X'),
                     '0001-2345-6789-012X');
    test.strictEqual(ORCID.toDashFormat('0000111122223336'),
                     '0000-1111-2222-3336');
    test.strictEqual(ORCID.toDashFormat('orcid.org/0000-1111-2222-3336'),
                     '0000-1111-2222-3336');
    test.strictEqual(ORCID.toDashFormat('orcid.org/0000111122223336'),
                     '0000-1111-2222-3336');
    test.strictEqual(ORCID.toDashFormat('https://orcid.org/0000-1111-2222-3336'),
                     '0000-1111-2222-3336');
    test.strictEqual(ORCID.toDashFormat('https://orcid.org/0000111122223336'),
                     '0000-1111-2222-3336');
    test.strictEqual(ORCID.toDashFormat('http://orcid.org/0000-1111-2222-3336'),
                     '0000-1111-2222-3336');
    test.strictEqual(ORCID.toDashFormat('http://orcid.org/0000111122223336'),
                     '0000-1111-2222-3336');
    test.done();
  },

  x_is_made_uppercase: function(test) {
    test.strictEqual(ORCID.toDashFormat('0001-2345-6789-012x'), '0001-2345-6789-012X');
    test.strictEqual(ORCID.toDashFormat('0001-2345-6789-012X'), '0001-2345-6789-012X');
    test.strictEqual(ORCID.toDashFormat('orcid.org/0001-2345-6789-012x'), '0001-2345-6789-012X');
    test.done();
  },

  invalid_ORCIDs_throw: function(test) {
    test.throws(function() {ORCID.toDashFormat('0000-0000-0000-0003');});
    test.throws(function() {ORCID.toDashFormat('0001-2345-6789-0122');});
    test.throws(function() {ORCID.toDashFormat('0000-1111-2222-3331');});
    test.throws(function() {ORCID.toDashFormat('9999-9999-9999-9990');});
    test.throws(function() {ORCID.toDashFormat('');});
    test.throws(function() {ORCID.toDashFormat('0');});
    test.throws(function() {ORCID.toDashFormat('not a valid ORCID');});
    test.throws(function() {ORCID.toDashFormat('https://orcid.org/profile');});
    test.throws(function() {ORCID.toDashFormat('ftp://0000-1111-2222-3336');});
    test.throws(function() {ORCID.toDashFormat('orcid.com/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toDashFormat('orcid.org://https/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toDashFormat('ftp://orcid.org/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toDashFormat('https:\\orcid.org/0000-1111-2222-3336');});
    test.done();
  }
};

exports.toNoDashFormat = {
  non_string_inputs_throw_TypeError: function(test) {
    test.throws(function() { ORCID.toNoDashFormat(10); });
    test.throws(function() { ORCID.toNoDashFormat({}); });
    test.throws(function() { ORCID.toNoDashFormat([]); });
    test.throws(function() { ORCID.toNoDashFormat(NaN); });
    test.throws(function() { ORCID.toNoDashFormat(null); });
    test.throws(function() { ORCID.toNoDashFormat(undefined); });
    test.throws(function() { ORCID.toNoDashFormat(new Error()); });
    test.throws(function() { ORCID.toNoDashFormat(function() {}); });
    test.throws(function() { ORCID.toNoDashFormat(Symbol()); });
    test.throws(function() { ORCID.toNoDashFormat(/./); });
    test.done();
  },

  valid_ORCIDs_are_converted_to_standalone_non_dashed_format: function(test) {
    test.strictEqual(ORCID.toNoDashFormat('0000-0000-0000-0001'),
                     '0000000000000001');
    test.strictEqual(ORCID.toNoDashFormat('0001-2345-6789-012X'),
                     '000123456789012X');
    test.strictEqual(ORCID.toNoDashFormat('0000111122223336'),
                     '0000111122223336');
    test.strictEqual(ORCID.toNoDashFormat('orcid.org/0000-1111-2222-3336'),
                     '0000111122223336');
    test.strictEqual(ORCID.toNoDashFormat('orcid.org/0000111122223336'),
                     '0000111122223336');
    test.strictEqual(ORCID.toNoDashFormat('https://orcid.org/0000-1111-2222-3336'),
                     '0000111122223336');
    test.strictEqual(ORCID.toNoDashFormat('https://orcid.org/0000111122223336'),
                     '0000111122223336');
    test.strictEqual(ORCID.toNoDashFormat('http://orcid.org/0000-1111-2222-3336'),
                     '0000111122223336');
    test.strictEqual(ORCID.toNoDashFormat('http://orcid.org/0000111122223336'),
                     '0000111122223336');
    test.done();
  },

  x_is_made_uppercase: function(test) {
    test.strictEqual(ORCID.toNoDashFormat('0001-2345-6789-012x'), '000123456789012X');
    test.strictEqual(ORCID.toNoDashFormat('0001-2345-6789-012X'), '000123456789012X');
    test.strictEqual(ORCID.toNoDashFormat('orcid.org/0001-2345-6789-012x'), '000123456789012X');
    test.done();
  },

  invalid_ORCIDs_throw: function(test) {
    test.throws(function() {ORCID.toNoDashFormat('0000-0000-0000-0003');});
    test.throws(function() {ORCID.toNoDashFormat('0001-2345-6789-0122');});
    test.throws(function() {ORCID.toNoDashFormat('0000-1111-2222-3331');});
    test.throws(function() {ORCID.toNoDashFormat('9999-9999-9999-9990');});
    test.throws(function() {ORCID.toNoDashFormat('');});
    test.throws(function() {ORCID.toNoDashFormat('0');});
    test.throws(function() {ORCID.toNoDashFormat('not a valid ORCID');});
    test.throws(function() {ORCID.toNoDashFormat('https://orcid.org/profile');});
    test.throws(function() {ORCID.toNoDashFormat('ftp://0000-1111-2222-3336');});
    test.throws(function() {ORCID.toNoDashFormat('orcid.com/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toNoDashFormat('orcid.org://https/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toNoDashFormat('ftp://orcid.org/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toNoDashFormat('https:\\orcid.org/0000-1111-2222-3336');});
    test.done();
  }
};

exports.toUriWithoutProtocol = {
  non_string_inputs_throw_TypeError: function(test) {
    test.throws(function() { ORCID.toUriWithoutProtocol(10); });
    test.throws(function() { ORCID.toUriWithoutProtocol({}); });
    test.throws(function() { ORCID.toUriWithoutProtocol([]); });
    test.throws(function() { ORCID.toUriWithoutProtocol(NaN); });
    test.throws(function() { ORCID.toUriWithoutProtocol(null); });
    test.throws(function() { ORCID.toUriWithoutProtocol(undefined); });
    test.throws(function() { ORCID.toUriWithoutProtocol(new Error()); });
    test.throws(function() { ORCID.toUriWithoutProtocol(function() {}); });
    test.throws(function() { ORCID.toUriWithoutProtocol(Symbol()); });
    test.throws(function() { ORCID.toUriWithoutProtocol(/./); });
    test.done();
  },

  valid_ORCIDs_are_converted_to_uri_without_protocol: function(test) {
    test.strictEqual(ORCID.toUriWithoutProtocol('0000-0000-0000-0001'),
                     'orcid.org/0000-0000-0000-0001');
    test.strictEqual(ORCID.toUriWithoutProtocol('0001-2345-6789-012X'),
                     'orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithoutProtocol('0000111122223336'),
                     'orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithoutProtocol('orcid.org/0000-1111-2222-3336'),
                     'orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithoutProtocol('orcid.org/0000111122223336'),
                     'orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithoutProtocol('https://orcid.org/0000-1111-2222-3336'),
                     'orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithoutProtocol('https://orcid.org/0000111122223336'),
                     'orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithoutProtocol('http://orcid.org/0000-1111-2222-3336'),
                     'orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithoutProtocol('http://orcid.org/0000111122223336'),
                     'orcid.org/0000-1111-2222-3336');
    test.done();
  },

  x_is_made_uppercase: function(test) {
    test.strictEqual(ORCID.toUriWithoutProtocol('0001-2345-6789-012x'),
                     'orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithoutProtocol('0001-2345-6789-012X')
        , 'orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithoutProtocol('orcid.org/0001-2345-6789-012x'),
                     'orcid.org/0001-2345-6789-012X');
    test.done();
  },

  invalid_ORCIDs_throw: function(test) {
    test.throws(function() {ORCID.toUriWithoutProtocol('0000-0000-0000-0003');});
    test.throws(function() {ORCID.toUriWithoutProtocol('0001-2345-6789-0122');});
    test.throws(function() {ORCID.toUriWithoutProtocol('0000-1111-2222-3331');});
    test.throws(function() {ORCID.toUriWithoutProtocol('9999-9999-9999-9990');});
    test.throws(function() {ORCID.toUriWithoutProtocol('');});
    test.throws(function() {ORCID.toUriWithoutProtocol('0');});
    test.throws(function() {ORCID.toUriWithoutProtocol('not a valid ORCID');});
    test.throws(function() {ORCID.toUriWithoutProtocol('https://orcid.org/profile');});
    test.throws(function() {ORCID.toUriWithoutProtocol('ftp://0000-1111-2222-3336');});
    test.throws(function() {ORCID.toUriWithoutProtocol('orcid.com/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toUriWithoutProtocol('orcid.org://https/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toUriWithoutProtocol('ftp://orcid.org/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toUriWithoutProtocol('https:\\orcid.org/0000-1111-2222-3336');});
    test.done();
  }
};

exports.toUriWithProtocol = {
  non_string_inputs_throw_TypeError: function(test) {
    test.throws(function() { ORCID.toUriWithProtocol(10); });
    test.throws(function() { ORCID.toUriWithProtocol({}); });
    test.throws(function() { ORCID.toUriWithProtocol([]); });
    test.throws(function() { ORCID.toUriWithProtocol(NaN); });
    test.throws(function() { ORCID.toUriWithProtocol(null); });
    test.throws(function() { ORCID.toUriWithProtocol(undefined); });
    test.throws(function() { ORCID.toUriWithProtocol(new Error()); });
    test.throws(function() { ORCID.toUriWithProtocol(function() {}); });
    test.throws(function() { ORCID.toUriWithProtocol(Symbol()); });
    test.throws(function() { ORCID.toUriWithProtocol(/./); });
    test.done();
  },

  non_boolean_and_non_undefined_flags_throw_TypeError: function(test) {
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', 10); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', {}); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', []); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', NaN); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', null); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', new Error()); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', function() {}); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', Symbol()); });
    test.throws(function() { ORCID.toUriWithProtocol('0000-0000-0000-0001', /./); });
    test.done();
  },

  omitting_the_secure_flag_defaults_to_secure_uri: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('0000-0000-0000-0001'),
                     'https://orcid.org/0000-0000-0000-0001');
    test.done();
  },

  setting_the_secure_flag_true_forces_https: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('0000-0000-0000-0001', true),
                     'https://orcid.org/0000-0000-0000-0001');
    test.done();
  },

  setting_the_secure_flag_flase_forces_http: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('0000-0000-0000-0001', false),
                     'http://orcid.org/0000-0000-0000-0001');
    test.done();
  },

  valid_ORCIDs_are_converted_to_uri_with_protocol: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('0000-0000-0000-0001'),
                     'https://orcid.org/0000-0000-0000-0001');
    test.strictEqual(ORCID.toUriWithProtocol('0001-2345-6789-012X'),
                     'https://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('0000111122223336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithProtocol('orcid.org/0000-1111-2222-3336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithProtocol('orcid.org/0000111122223336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithProtocol('https://orcid.org/0000-1111-2222-3336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithProtocol('https://orcid.org/0000111122223336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithProtocol('http://orcid.org/0000-1111-2222-3336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.strictEqual(ORCID.toUriWithProtocol('http://orcid.org/0000111122223336'),
                     'https://orcid.org/0000-1111-2222-3336');
    test.done();
  },

  x_is_made_uppercase: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('0001-2345-6789-012x'),
                     'https://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('0001-2345-6789-012X'),
                     'https://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('orcid.org/0001-2345-6789-012x'),
                     'https://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('0001-2345-6789-012x', false),
                     'http://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('0001-2345-6789-012X', false),
                     'http://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('orcid.org/0001-2345-6789-012x', false),
                     'http://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('http://orcid.org/0001-2345-6789-012x', false),
                     'http://orcid.org/0001-2345-6789-012X');
    test.done();
  },

  protocol_of_input_can_be_changed_using_flag: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('http://orcid.org/0001-2345-6789-012X', true),
                     'https://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('https://orcid.org/0001-2345-6789-012X', false),
                     'http://orcid.org/0001-2345-6789-012X');
    test.done();
  },
  protocol_gets_upgraded_when_flag_is_omitted: function(test) {
    test.strictEqual(ORCID.toUriWithProtocol('https://orcid.org/0001-2345-6789-012X'),
                     'https://orcid.org/0001-2345-6789-012X');
    test.strictEqual(ORCID.toUriWithProtocol('https://orcid.org/0001-2345-6789-012X'),
                     'https://orcid.org/0001-2345-6789-012X');
    test.done();
  },

  invalid_ORCIDs_throw: function(test) {
    test.throws(function() {ORCID.toUriWithProtocol('0000-0000-0000-0003');});
    test.throws(function() {ORCID.toUriWithProtocol('0001-2345-6789-0122');});
    test.throws(function() {ORCID.toUriWithProtocol('0000-1111-2222-3331', true);});
    test.throws(function() {ORCID.toUriWithProtocol('9999-9999-9999-9990', false);});
    test.throws(function() {ORCID.toUriWithProtocol('');});
    test.throws(function() {ORCID.toUriWithProtocol('0', true);});
    test.throws(function() {ORCID.toUriWithProtocol('not a valid ORCID', false);});
    test.throws(function() {ORCID.toUriWithProtocol('https://orcid.org/profile');});
    test.throws(function() {ORCID.toUriWithProtocol('ftp://0000-1111-2222-3336', true);});
    test.throws(function() {ORCID.toUriWithProtocol('orcid.com/0000-1111-2222-3336', false);});
    test.throws(function() {ORCID.toUriWithProtocol('orcid.org://https/0000-1111-2222-3336');});
    test.throws(function() {ORCID.toUriWithProtocol('ftp://orcid.org/0000-1111-2222-3336', true);});
    test.throws(
        function() {ORCID.toUriWithProtocol('https:\\orcid.org/0000-1111-2222-3336', false);});
    test.done();
  }
};

exports.isValid = {
  non_string_inputs_throw_TypeError: function(test) {
    test.throws(function() { ORCID.isValid(10); });
    test.throws(function() { ORCID.isValid({}); });
    test.throws(function() { ORCID.isValid([]); });
    test.throws(function() { ORCID.isValid(NaN); });
    test.throws(function() { ORCID.isValid(null); });
    test.throws(function() { ORCID.isValid(undefined); });
    test.throws(function() { ORCID.isValid(new Error()); });
    test.throws(function() { ORCID.isValid(function() {}); });
    test.throws(function() { ORCID.isValid(Symbol()); });
    test.throws(function() { ORCID.isValid(/./); });
    test.done();
  },

  known_good_ORCIDs_are_valid: function(test) {
    test.strictEqual(ORCID.isValid('0000-0000-0000-0001'), true);
    test.strictEqual(ORCID.isValid('0001-2345-6789-012X'), true);
    test.strictEqual(ORCID.isValid('0000-1111-2222-3336'), true);
    test.strictEqual(ORCID.isValid('9999-9999-9999-9994'), true);
    test.done();
  },

  valid_ORCIDs_can_be_in_any_acceptable_format: function(test) {
    test.strictEqual(ORCID.isValid('0000-1111-2222-3336'), true);
    test.strictEqual(ORCID.isValid('0000111122223336'), true);
    test.strictEqual(ORCID.isValid('orcid.org/0000-1111-2222-3336'), true);
    test.strictEqual(ORCID.isValid('orcid.org/0000111122223336'), true);
    test.strictEqual(ORCID.isValid('https://orcid.org/0000-1111-2222-3336'), true);
    test.strictEqual(ORCID.isValid('https://orcid.org/0000111122223336'), true);
    test.strictEqual(ORCID.isValid('http://orcid.org/0000-1111-2222-3336'), true);
    test.strictEqual(ORCID.isValid('http://orcid.org/0000111122223336'), true);
    test.done();
  },

  known_bad_ORCIDs_are_not_valid: function(test) {
    test.strictEqual(ORCID.isValid('0000-0000-0000-0003'), false);
    test.strictEqual(ORCID.isValid('0001-2345-6789-0122'), false);
    test.strictEqual(ORCID.isValid('0000-1111-2222-3331'), false);
    test.strictEqual(ORCID.isValid('9999-9999-9999-9990'), false);
    test.done();
  },

  strings_without_ORCIDs_in_are_not_valid: function(test) {
    test.strictEqual(ORCID.isValid(''), false);
    test.strictEqual(ORCID.isValid('0'), false);
    test.strictEqual(ORCID.isValid('not a valid ORCID'), false);
    test.strictEqual(ORCID.isValid('https://orcid.org/profile'), false);
    test.done();
  },

  unrecognised_prefixes_throw_Error: function(test) {
    test.throws(ORCID.isValid('ftp://0000-1111-2222-3336'));
    test.throws(ORCID.isValid('orcid.com/0000-1111-2222-3336'));
    test.throws(ORCID.isValid('orcid.org://https/0000-1111-2222-3336'));
    test.throws(ORCID.isValid('ftp://orcid.org/0000-1111-2222-3336'));
    test.throws(ORCID.isValid('https:\\orcid.org/0000-1111-2222-3336'));
    test.done();
  },

  prefix_ordering_must_have_protocol_first: function(test) {
    test.throws(ORCID.isValid('orcid.org/https://0000-1111-2222-3336'));
    test.ok(ORCID.isValid('https://orcid.org/0000-1111-2222-3336'));
    test.done();
  },

  ORCIDs_can_be_case_insensitive: function(test) {
    test.strictEqual(ORCID.isValid('0000-0000-0000-001X'),
                     ORCID.isValid('0000-0000-0000-001x'));
    test.strictEqual(ORCID.isValid('orcid.org/0000-1111-2222-3336'),
                     ORCID.isValid('ORCID.ORG/0000-1111-2222-3336'));
    test.strictEqual(ORCID.isValid('https://orcid.org/0000-1111-2222-3336'),
                     ORCID.isValid('https://ORCID.ORG/0000-1111-2222-3336'));
    test.strictEqual(ORCID.isValid('https://orcid.org/0000-1111-2222-3336'),
                     ORCID.isValid('HTTPS://ORCID.ORG/0000-1111-2222-3336'));
    test.done();
  }
};


src/orcid.ts
'use strict'

/**
 * Checks that a string is a valid ORCID of any format
 * @param {string} input String to test format of
 * @returns {boolean}
 */
function inAcceptedFormat(input: string): boolean {
  if (typeof input !== 'string') throw TypeError('Input must be string')
  return /^((https?:\/\/)?orcid\.org\/)?([0-9]{4}-?){3}[0-9]{3}[0-9X]$/i.test(input)
}

/**
 * Converts a valid ORCID (of any format) into the dashed format
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toDashFormat(input: string): string {
  const noDash = toNoDashFormat(input)
  return [0, 4, 8, 12].map(it => noDash.slice(it, it + 4)).join('-')
}

/**
 * Converts a valid ORCID (of any format) into the non-dashed format (without validating the input first - useful internally)
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toNoDashFormatWithoutValidation(input: string): string {
  return input.replace(/-/g, '').slice(-16).toUpperCase()
}

/**
 * Converts a valid ORCID (of any format) into the non-dashed format
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toNoDashFormat(input: string): string {
  validate(input)
  return toNoDashFormatWithoutValidation(input)
}

/**
 * Converts an ORCID to its corresponding URI, e.g. https://orcid.org/xxxxxxx......
 * @param {string} input Any valid ORCID
 * @param {boolean=} secure true (default): https, false: http
 * @returns {string}
 */
function toUriWithProtocol(input: string, secure?: boolean): string {
  return (secure === undefined || !!secure ? 'https' : 'http') + '://' + toUriWithoutProtocol(input)
}

/**
 * Converts an ORCID to is corresponding URI, without a protocol, e.g. orcid.org/xxxxxx.....
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toUriWithoutProtocol(input: string): string {
  return 'orcid.org/' + toDashFormat(input)
}

/**
 * Returns true or false indicating whether a string conforms to any format of ORCID.
 * @param {string} input Any string to test
 * @returns {boolean}
 */
function isValid(input: string): boolean {
  if (!inAcceptedFormat(input)) return false
  const inputNoDash = toNoDashFormatWithoutValidation(input)

  let total = 0
  for (let i = 0; i < 15; i++) {
    total = (total + +inputNoDash[i]) * 2
  }
  const result = (12 - (total % 11)) % 11
  return inputNoDash[15] === (result === 10 ? 'X' : result + '')
}

/**
 * A side-effect version of `isValid` - will throw Error if `isValid` would return false.
 * @param {string} input
 * @returns {void}
 */
function validate(input: string): void {
  if (!isValid(input)) throw Error('Invalid ORCID')
}

const ORCID = {
  isValid,
  validate,
  inAcceptedFormat,
  toDashFormat,
  toNoDashFormat,
  toUriWithProtocol,
  toUriWithoutProtocol,
}
if (typeof exports !== 'undefined') Object.assign(exports, ORCID)
if (typeof window !== 'undefined') Object.assign(window, { ORCID })


lib/orcid.d.ts
/**
 * Checks that a string is a valid ORCID of any format
 * @param {string} input String to test format of
 * @returns {boolean}
 */
declare function inAcceptedFormat(input: string): boolean;
/**
 * Converts a valid ORCID (of any format) into the dashed format
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
declare function toDashFormat(input: string): string;
/**
 * Converts a valid ORCID (of any format) into the non-dashed format (without validating the input first - useful internally)
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
declare function toNoDashFormatWithoutValidation(input: string): string;
/**
 * Converts a valid ORCID (of any format) into the non-dashed format
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
declare function toNoDashFormat(input: string): string;
/**
 * Converts an ORCID to its corresponding URI, e.g. https://orcid.org/xxxxxxx......
 * @param {string} input Any valid ORCID
 * @param {boolean=} secure true (default): https, false: http
 * @returns {string}
 */
declare function toUriWithProtocol(input: string, secure?: boolean): string;
/**
 * Converts an ORCID to is corresponding URI, without a protocol, e.g. orcid.org/xxxxxx.....
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
declare function toUriWithoutProtocol(input: string): string;
/**
 * Returns true or false indicating whether a string conforms to any format of ORCID.
 * @param {string} input Any string to test
 * @returns {boolean}
 */
declare function isValid(input: string): boolean;
/**
 * A side-effect version of `isValid` - will throw Error if `isValid` would return false.
 * @param {string} input
 * @returns {void}
 */
declare function validate(input: string): void;
declare const ORCID: {
    isValid: typeof isValid;
    validate: typeof validate;
    inAcceptedFormat: typeof inAcceptedFormat;
    toDashFormat: typeof toDashFormat;
    toNoDashFormat: typeof toNoDashFormat;
    toUriWithProtocol: typeof toUriWithProtocol;
    toUriWithoutProtocol: typeof toUriWithoutProtocol;
};
//# sourceMappingURL=orcid.d.ts.map

lib/orcid.js
'use strict'

/**
 * Checks that a string is a valid ORCID of any format
 * @param {string} input String to test format of
 * @returns {boolean}
 */
function inAcceptedFormat(input) {
  if (typeof input !== 'string') throw TypeError('Input must be string')
  return /^((https?:\/\/)?orcid\.org\/)?([0-9]{4}-?){3}[0-9]{3}[0-9X]$/i.test(input)
}

/**
 * Converts a valid ORCID (of any format) into the dashed format
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toDashFormat(input) {
  const noDash = toNoDashFormat(input)
  return [0, 4, 8, 12].map(it => noDash.slice(it, it + 4)).join('-')
}

/**
 * Converts a valid ORCID (of any format) into the non-dashed format (without validating the input first - useful internally)
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toNoDashFormatWithoutValidation(input) {
  return input.replace(/-/g, '').slice(-16).toUpperCase()
}

/**
 * Converts a valid ORCID (of any format) into the non-dashed format
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toNoDashFormat(input) {
  validate(input)
  return toNoDashFormatWithoutValidation(input)
}

/**
 * Converts an ORCID to its corresponding URI, e.g. https://orcid.org/xxxxxxx......
 * @param {string} input Any valid ORCID
 * @param {boolean=} secure true (default): https, false: http
 * @returns {string}
 */
function toUriWithProtocol(input, secure) {
  const proto = { true: 'https', false: 'http' }[secure === undefined || !!secure]
  return proto + '://' + toUriWithoutProtocol(input)
}

/**
 * Converts an ORCID to is corresponding URI, without a protocol, e.g. orcid.org/xxxxxx.....
 * @param {string} input Any valid ORCID
 * @returns {string}
 */
function toUriWithoutProtocol(input) {
  return 'orcid.org/' + toDashFormat(input)
}

/**
 * Returns true or false indicating whether a string conforms to any format of ORCID.
 * @param {string} input Any string to test
 * @returns {boolean}
 */
function isValid(input) {
  if (!inAcceptedFormat(input)) return false
  const inputNoDash = toNoDashFormatWithoutValidation(input)

  let total = 0
  for (let i = 0; i < 15; i++) {
    total = (total + +inputNoDash[i]) * 2
  }
  const result = (12 - (total % 11)) % 11
  return inputNoDash[15] === (result === 10 ? 'X' : result + '')
}

/**
 * A side-effect version of `isValid` - will throw Error if `isValid` would return false.
 * @param {string} input
 * @returns {void}
 */
function validate(input) {
  if (!isValid(input)) throw Error('Invalid ORCID')
}

const ORCID = {
  isValid,
  validate,
  inAcceptedFormat,
  toDashFormat,
  toNoDashFormat,
  toUriWithProtocol,
  toUriWithoutProtocol,
}
if (typeof exports !== 'undefined') Object.assign(exports, ORCID)
if (typeof window !== 'undefined') window['ORCID'] = ORCID
